### PR TITLE
Extension of the `ReportPortal` plugin using API

### PR DESCRIPTION
**Pull request by 4N0body5**
**PR should cover following use cases:**

1.|  Create one launch for all plans and create suite per each plan
2.|  Additionaly upload new suite/tests to a an existing launch
3.|  Prepare launch with empty (idle) tests and additionally report their results (within the same run id)
4.|  Rerun and upload new version into new `Retry` item within the last launch based on mapping via names.
5.|  Upload files
6.|  Enable using v2

**New implemented options:**

* option `--launch-per-plan` for creating a new launch per each plan (default)
* option `--suite-per-plan` for mapping a suite per each plan, all enclosed in one launch (launch uuid is stored in run of the first plan)
* option `--launch-description` for providing unified launch description, intended mainly for suite-per-plan mapping
* option `--upload-to-launch LAUNCH_ID`  to append new plans to an existing launch
* option `--upload-to-suite SUITE_ID` to append new tests to an existing suite within launch
* option `--launch-rerun` for reruns with 'Retry' item in RP, name-based mapping, only for suite-per-plan structure.
* option `--defect-type` for passing the defect type to failing tests, enables report idle tests to be additionally updated. 

**Additional work behind that:**

* rewritten environment variables to the uniform form TMT_PLUGIN_REPORT_REPORTPORTAL_${option}
* enabled to report without results
* get launch attributes from all common ones in plans (suite-per-plan)
* storing uuids per each launch/suite followed by reusing them (update based on identifiers)
* analyzed option combinations
* version v2

**Yet todo:**

* uploading to suites
* check stored identifiers
* restrict problematic option combinations
* resolve problems with mypy
* and lot of small debugging
* schemas, documentation
* tests (for new PR)
* uploading files (for new PR)

### CODE REVIEW

File: tmt/steps/report/reportportal.py [hunk 1]
```diff
@@ -15,12 +15,12 @@ class ReportReportPortalData(tmt.steps.report.ReportStepData):
...
         option="--url",
         metavar="URL",
-        default=os.environ.get("TMT_REPORT_REPORTPORTAL_URL"),
+        default=None,
         help="The URL of the ReportPortal instance where the data should be sent to.")
     token: Optional[str] = field(
         option="--token",
         metavar="TOKEN",
-        default=os.environ.get("TMT_REPORT_REPORTPORTAL_TOKEN"),
+        default=None,
```
Comment (happz) [inline tmt/steps/report/reportportal.py:23]: IIUIC, the `os.environ` is necessary. See the note at https://tmt.readthedocs.io/en/stable/overview.html#plugin-variables: tmt will extract field value from environment variable only when the step and plugin are mentioned on the command line. This means, in the general case of TF, with the plugin listed in fmf files only, the environment variables would be ignored and have no effect. `os.environ` is a workaround: when executed, the `default` is filled with whatever is in the environment, tricking tmt into honoring the environment after all.

Comment (happz) [reply tmt/steps/report/reportportal.py:23]: I noticed you moved `os.environ` use into the code - I'd recommend against it, as it splits field value initialization into several places. One day, environment variables would be honored even without listing the plugin on the command line, `os.environ` trick wouldn't be necessary, and hunting down all those suddenly useless `os.environ.get()` calls would be easier should they happen in step data classes only.

File: tmt/base.py [hunk 2]
```diff
@@ -2962,6 +2962,7 @@ class RunData(tmt.utils.SerializableContainer):
     steps: List[str]
     environment: EnvironmentType
     remove: bool
+    rp_uuid: Optional[str]
```
Comment (happz) [inline tmt/base.py:2965]: Does it work? :)

If it does, we should make it more generic. I'd like to avoid ad hoc additions like this, `rp` can mean many things, if you decide you need to store 10 more attributes, we'd have 11 `rp_*` fields in this class, etc.

I'd go with some classes right from the beginning, but `RunData` is not ready, so let's start with dictionaries. How about something like this: `plugin_data: Dict[str, Dict[str, str]]`? The first key would be the plugin name, the nested dictionary would be owned by the plugin, store there whatever you wish. I expect something like this would materialize:

```python
plugin_data: {
  'reportportal': {
    'uuid': ...
  }
}
```

I'd introduce dataclasses later, we can do it in a standalone patch.

File: tmt/steps/report/reportportal.py (hunk 1 above)
Comment (4N0body5) [reply tmt/steps/report/reportportal.py:23]: I followed the implementation of polarion plugin. My intention was mostly to change it into TMT_PLUGIN_REPORT_REPORTPORTAL_<option> form (instead of TMT_REPORT_REPORTPORTAL_<token/url>) and have it uniform.

File: tmt/base.py (hunk 2 above)
Comment (4N0body5) [reply tmt/base.py:2965]: Sure, it was temporary, I left it open to suggestions, perhaps on the following meeting. But I think your idea looks good, so I can redesign it that way.

**Comment by brakogh** [comment]
@4N0body5 can you share progress pls?

**Comment by psss** [comment]
@4N0body5, is the `WIP` tag still relevant or is the pull request ready for review?

**Comment by 4N0body5** [comment]
> @4N0body5, is the `WIP` tag still relevant or is the pull request ready for review?

@psss  This week, I plan to look into uploading files, setting version, fixing the problem with uploading to suites, and resolving problems due to option combinations and plan to remove the tag afterwards.

**Comment by psss** [comment]
Thanks for the update, @4N0body5. Nice progress!

> **PR should cover following use cases:**
> 
> 1. Create one launch for all plans and create suite per each plan
> 2. Additionaly upload new suite/tests to a an existing launch
> 3. Prepare launch with empty (idle) tests and additionally report their results (within the same run id)
> 4. Rerun and upload new version into new `Retry` item within the last launch based on mapping via names.
> 5. Upload files
> 6. Enable using v2
> 
> **New implemented options:**
> 
> * option `--launch-per-plan` for creating a new launch per each plan (default)
> * option `--suite-per-plan` for mapping a suite per each plan, all enclosed in one launch (launch uuid is stored in run of the first plan)
> * option `--launch-description` for providing unified launch description, intended mainly for suite-per-plan mapping
> * option `--upload-to-launch LAUNCH_ID`  to append new plans to an existing launch
> * option `--upload-to-suite SUITE_ID` to append new tests to an existing suite within launch
> * option `--launch-rerun` for reruns with 'Retry' item in RP, name-based mapping, only for suite-per-plan structure.
> * option `--defect-type` for passing the defect type to failing tests, enables report idle tests to be additionally updated.

Sounds like a large bunch of new features! Would it make sense to cover e.g. the upload files in a separate pull request and not block already finished changes on it? Or is everything here too much intertwined that it would not make sense to have smaller pull requests dedicated to a single change / smaller scope?

**Comment by 4N0body5** [comment]
> ... Would it make sense to cover e.g. the upload files in a separate pull request and not block already finished changes on it? Or is everything here too much intertwined that it would not make sense to have smaller pull requests dedicated to a single change / smaller scope?

@psss It does make sense. I guess I'll make a brief look into the issue; If it is no struggle, I'll try adding it, so it can be effectively reviewed with all of others, but as soon as it shows any sign of difficulties, I'll suggest postponing it into another PR.
E.g. with the version I only hope to add one environment variable as a quick (maybe temporary) solution, but I'm not so sure about the upload files.
Definitely, the main priority is debug those already implemented parts.

File: tmt/base.py (hunk 2 above)
Comment (happz) [reply tmt/base.py:3092]: True. Just check whether the `report.data[0]` is the same report plugin, to avoid surprises. E.g. `rp_phases = [phase for phase in self.step.plan.my_run.plans[0].report.phases(ReportReportPortal)]` should give you their list - if it's empty, there was no RP report in the plan, otherwise you can pick the first item of this list and be sure it does have `launch_uuid`.

File: tmt/steps/report/reportportal.py [hunk 3]
```diff
@@ -91,77 +139,222 @@ class ReportReportPortal(tmt.steps.report.ReportPlugin[ReportReportPortalData]):
...
         """
 
         if not response.ok:
             raise tmt.utils.ReportError(
                 f"Received non-ok status code from ReportPortal: {response.text}")
 
-        self.debug("Response code from the server", response.status_code)
-        self.debug("Message from the server", response.text)
+        self.debug("Response code from the endpoint", str(response.status_code))
+        self.debug("Message from the endpoint", str(response.text))
```
Comment (happz) [inline tmt/steps/report/reportportal.py:150]: Neither of these `str()` calls should be needed, recent changes in `debug()` now support `int` values out of the box (`status_code`), and `text` is already a string.

File: tmt/steps/report/reportportal.py [hunk 4]
```diff
@@ -15,49 +16,95 @@ class ReportReportPortalData(tmt.steps.report.ReportStepData):
...
-        help="""
-             Regular expression for excluding environment variables from reporting to ReportPortal
-             ('^TMT_.*' used by default).
-             """)
+        help="Regular expression for excluding environment variables "
+             "from reporting to ReportPortal ('^TMT_.*' used by default).")
+    launch_url: str = ""
+    launch_uuid: str = ""
+    suite_uuid: str = ""
+    test_uuids: list[str] = ""
```
Comment (happz) [inline tmt/steps/report/reportportal.py:89]: Four weird-looking fields :)

* unless an empty string is really a valid launch URL/launch UUID/suite UUID, these three should be `Optional[str] = None`. `None` serves better as an expression of "not set", way better than an empty string which is much harder to follow for linters.
* `list[str] = ""` will not be accepted by linter, and either the type is incorrect, or the initial value is.

File: tmt/steps/report/reportportal.py [hunk 5]
```diff
@@ -91,77 +139,222 @@ class ReportReportPortal(tmt.steps.report.ReportPlugin[ReportReportPortalData]):
...
+        #                   - forbid upload-to-suite && launch-per-plan
+        #       * try read a launch_uuid at first plan (self.step.plan.report.launch_uuid)
+
         super().go()
 
-        server = self.data.url
-        if not server:
-            raise tmt.utils.ReportError("No ReportPortal server url provided.")
-        server = server.rstrip("/")
+        endpoint = self.get("url")
```
Comment (happz) [inline tmt/steps/report/reportportal.py:183]: I suspect this is a combination of the PR being fairly old and changes that happened in the meantime, rather than the actual change you'd like to make. But, if that's an actual change, I'd strongly advise not to use `get()`, `self.data` should be used instead. `self.data.foo` is properly annotated while `self.get('foo')` returns `Any`, lowering guarantees we get from type annotation linters.

File: tmt/steps/report/reportportal.py [hunk 6]
```diff
@@ -91,77 +139,222 @@ class ReportReportPortal(tmt.steps.report.ReportPlugin[ReportReportPortalData]):
...
             "Authorization": "bearer " + token,
             "accept": "*/*",
             "Content-Type": "application/json"}
 
-        envar_pattern = self.data.exclude_variables
+        launch_time = str(int(time() * 1000))
+
+        # to support idle tests
+        executed = False
+        if len(self.step.plan.execute.results()) > 0:
```
Comment (happz) [inline tmt/steps/report/reportportal.py:206]: `if self.step.plan.execute.results():` is more Pythonic - empty list would be false-ish.

```python
executed = bool(self.step.plan.execute.results())
if executed:
    launch_time = self.step.plan.execute.results()[0].start_time
```

File: tmt/base.py (hunk 2 above)
Comment (happz) [reply tmt/base.py:3092]: I suppose the field can be removed from the PR, correct?

File: tmt/steps/report/reportportal.py [hunk 7]
```diff
@@ -91,77 +139,222 @@ class ReportReportPortal(tmt.steps.report.ReportPlugin[ReportReportPortalData]):
...
+                tmp_dict = {}
+                for key, value in current_plan.items():
+                    if key in result_dict and result_dict[key] == value:
+                        tmp_dict[key] = value
+                result_dict = tmp_dict
+            launch_attributes = [
+                {'key': key, 'value': value}
+                for key, value in result_dict.items()]
+        else:
+            launch_attributes = attributes.copy()
```
Comment (happz) [inline tmt/steps/report/reportportal.py:256]: Something to consider, maybe: I'm worried that the resulting `go()` method will be long and hard to follow and maintain. Would it make sense to split it (or most of it) into smaller functions, working together, into helper functions designed for one particular step of the bigger picture?

E.g. the `if suite_per_plan` above. It might be turned into a nifty function, something like the following:

```python
def construct_launch_attributes(self, suite_per_plan: bool, attributes: Dict[str, str]) -> Dict[str, str]:
    if not suite_per_plan:
        return attributes.copy()

    merged_plans = [{key: value[0] for key, value in plan._fmf_context.items()}
                    for plan in self.step.plan.my_run.plans]
    result_dict = merged_plans[0]
    ...
```

In this method, only `launch_atrributes = self.construct_launch_attributes(suite_per_plan, attributes)` would remain. `go()` would be easier to follow, dealing with logical steps on a higher level, `construct_launch_attributes` would be much easier to test with unit tests as it would define its inputs and return value - you cannot reasonably test the creation of launch attributes now -, and `go()` wouldn't be full of various one-shot variables.

And there are multiple such blocks that could be isolated, e.g. `dt_locator` - take the code under `if detect_type`, make it fetch and return the defect type, and `dt_locator = self.get_defect_type(...) if detect_type else None`.

I don't know how far in your work you are, but I'm worried we're getting a bit too close to a second `DiscoverFmf.go()`.

@psss what do you think?

Comment (psss) [reply tmt/steps/report/reportportal.py:256]: > Something to consider, maybe: I'm worried that the resulting `go()` method will be long and hard to follow and maintain. Would it make sense to split it (or most of it) into smaller functions, working together, into helper functions designed for one particular step of the bigger picture?

Yes, agreed, this would definitely help with readability and maintainability as well, +1 from me.

**Comment by kkaarreell** [comment]
One thing I am struggling with is how to properly use options like `--launch-rerun`, `--upload-to-suite` etc for a rerun. An example would be nice.

**Comment by kkaarreell** [comment]
One more thing to think about. How a user or an automation will find out URLs od created RP launches. Especially once tests will ve running through TF.

**Comment by 4N0body5** [comment]
> One more thing to think about. How a user or an automation will find out URLs of created RP launches. Especially once tests will ve running through TF.

AFAIK I could only provide it via info log that lists the link in terminal. I'm not sure how TF works, let me know if there's any idea to make the url more available.

File: tmt/base.py (hunk 2 above)
Comment (thrix) [reply tmt/base.py]: @4N0body5 is this resolved?

File: tmt/steps/report/reportportal.py (hunk 6 above)
Comment (thrix) [reply tmt/steps/report/reportportal.py]: @4N0body5 is this resolved?

File: tmt/steps/report/reportportal.py (hunk 7 above)
Comment (thrix) [reply tmt/steps/report/reportportal.py]: @4N0body5 is this resolved?

Comment (happz) [reply tmt/steps/report/reportportal.py]: It's not, and it's probably too late now to start refactoring...

**Review by lukaszachy — APPROVED** [review]

**Review by happz — APPROVED** [review]
